pubsub/client.go

```diff
@@ -465,13 +279,76 @@ func (c *Client) GetSchema(schemaID string) (*eventbusv1.SchemaInfo, error) {
 		Fields(trailer).
 		Msg("retrieved schema")
 	if err != nil {
-		return nil, fmt.Errorf("error getting schema from salesforce api - %s", err)
+		return nil, errors.Errorf("error getting schema from salesforce api - %s", err)
 	}
 
 	return resp, nil
 }
 
-// Wrapper function around the Subscribe RPC. This will add the OAuth credentials and create a separate streaming client that will be used to,
+func (c *Client) Publish(ctx context.Context, records []opencdc.Record) error {
+	var events []*eventbusv1.ProducerEvent
+	// TODO - refactor the way we access the topic
+	topic, err := c.GetTopic(c.topicNames[0])
+	if err != nil {
+		return errors.Errorf("error on publish, cannot retrieve topic %s : %s", c.topicNames[0], err)
+	}
+
+	sdk.Logger(ctx).Info().
+		Str("topic", topic.TopicName).
+		Msg("Started publishing event")
+
+	codec, err := c.fetchCodec(ctx, topic.SchemaId)
+	if err != nil {
+		return err
+	}
+
+	for _, r := range records {
+		data, err := extractPayload(r.Operation, r.Payload)
+		if err != nil {
+			return errors.Errorf("failed to extract payload data: %s", err)
+		}
+		avroPrepared, err := validateAndPreparePayload(data, codec.Schema())
+		if err != nil {
+			return errors.Errorf("error validating and preparing avro data:%s", err)
+		}
+
+		avroEncoded, err := codec.BinaryFromNative(nil, avroPrepared)
+		if err != nil {
+			return errors.Errorf("error encoding data to avro: %s", err)
+		}
+
+		event := eventbusv1.ProducerEvent{
+			SchemaId: topic.SchemaId,
+			Payload:  avroEncoded,
+		}
+		events = append(events, &event)
+	}
+
+	publishRequest := eventbusv1.PublishRequest{
```

Comment:
Is it possible that this method would publish some events and then fail, so some are published and some are not?

Reply:
I did some looking into, and it seems like its possible it might happen :( Good call out! 
What do you think the behavior for it should be like? We can retry publish n amount of times for the failed events, does that makes sense for a connector to do?

Reply:
if you know what events failed and can retry them, then sure why not! the destination "Write" method should return the number of records that were successfully sent, so we know what was written and what was not.